### ISSUE TITLE
Edge DevTools VSCode Extension not working on VSCode Insiders

### ISSUE THREAD

**Issue by mliao95** [microsoft/vscode#101562]
- VSCode Version: 1.47.0 (vscode-insiders)
- OS Version: Win 64

Steps to Reproduce:
1.  Set up the VSCode DevTools Extension ([Instructions](https://github.com/microsoft/vscode-edge-devtools/blob/master/CONTRIBUTING.md))
2.  Run in VSCode Insiders

Does this issue occur when all extensions are disabled?:  Not applicable

Hey!  I'm working on the VSCode Edge DevTools extension, and noticed that our extension no longer works on vscode insiders, but works properly on the stable version.  I was hoping we could find a point of contact on the vscode team to help us work through what compatibility issues might have come up between stable and the insiders version.

For a bit of reference here is how we serve our DevTools via the extension:
![image](https://user-images.githubusercontent.com/14304598/86311586-9296b580-bbd5-11ea-8653-9d06237f3950.png)

I believe that the compatibility issue occurs somewhere between the Host and the DevTools as the main issue is that our DevTools files is not reaching the host.  The main file that deals with serving the DevTools is [host.ts](https://github.com/microsoft/vscode-edge-devtools/blob/master/src/host/host.ts).  

Expected:
Application runs as normal
![image](https://user-images.githubusercontent.com/14304598/86311476-5a8f7280-bbd5-11ea-9360-dd5ee8ff3b27.png)

Actual:
Blank EdgeDevTools tab:
![image](https://user-images.githubusercontent.com/14304598/86311364-156b4080-bbd5-11ea-925e-d5e9fbaa41bc.png)

**Comment by alexdima** [comment]
It looks like the WebView can no longer communicate to Edge to load the developer tools, so I think this might have been caused by some changes to the web view area.

**Comment by mjbvz** [comment]
@mliao95 Can you please provide steps to reproduce this? I cannot get the edge dev tools to launch/attach consistently enough to debug this. It worked maybe once out of ten tries:

![Screen Shot 2020-07-02 at 11 45 59 AM](https://user-images.githubusercontent.com/12821956/86398037-cddac800-bc59-11ea-9a59-fbdc3057efd7.png)

**Comment by mjbvz** [comment]
Also which insiders build are you testing on?

**Comment by mjbvz** [comment]
I also tried building from source on windows but ran into the same issue. All I see in the attach targets are our internal VS Code instances:

<img width="960" alt="Screen Shot 2020-07-02 at 2 12 31 PM" src="https://user-images.githubusercontent.com/12821956/86409616-1ac89980-bc6e-11ea-8126-a2a79be44967.png">

**Comment by mjbvz** [comment]
Ok, so I can reproduce this by attaching to one of those targets but can't figure out what is going wrong without more context on how the extension is implemented

@mliao95 Can you please try debugging to see what specifically is not working in VS Code insiders?

**Comment by mliao95** [comment]
Hey Matt - sorry for the late reply!  I was out on Thursday, but I'll look into this tomorrow and get back to you.

**Comment by mliao95** [comment]
It appears that the main issue is that the `stylesUri` and `scriptUri` are not being loaded in properly in our `getHtmlForWebview()` function.  [Here ](https://github.com/microsoft/vscode-edge-devtools/blob/master/src/devtoolsPanel.ts#L274)is a link to the code.  Our hunch is that this might be due to the Content Security Policy blocking these scripts.

Also, this issue might be related to this existing Webview issue: https://github.com/microsoft/vscode/issues/101272

**Comment by mjbvz** [comment]
Here's a screenshot of what I see when the frame fails to load (I've set the undocumented `"webview.experimental.useIframes": true,` setting to make debugging this easier):

<img width="1481" alt="Screen Shot 2020-07-07 at 2 38 01 PM" src="https://user-images.githubusercontent.com/12821956/86846258-7e343c00-c05f-11ea-8465-ea4ea3817fdd.png">

It looks like all the top level resources are being loaded. There are a bunch of console error which seem to come from the extension. @mliao95 Would the `Incompatible embedder` errors explain the issue?

**Comment by mliao95** [comment]
I don't believe those are related.  We do not implement all the `Host.InspectorFrontendHost` functions, so a lot of them use the default stub.
![image](https://user-images.githubusercontent.com/14304598/86851478-e38c2b00-c067-11ea-8817-b05ba5510f28.png)

I forgot to add to the earlier post, but the other engineer on the extension, Vidal, managed to get the DevTools to render somewhat.  You can view the changes [here](https://github.com/microsoft/vscode-edge-devtools/commit/31d6c8908ba7be70ecb6b8f1058abc6dada387c3) or checkout the branch `vidal/webviewIssuesDraft`
![image](https://user-images.githubusercontent.com/14304598/86851736-51385700-c068-11ea-9b46-42a1a5c63f24.png)

We believe that this is related to the `stylesUri`/`scriptUri` not being loaded properly as stated in my previous post.  Have you seen this type of issue before where a styles script was not being loaded properly.

**Comment by mjbvz** [comment]
Oh yes those changes would explain part of this. This iteration we changed from loading resources using `vscode-resource://` to `vscode-webview-resource://` and also changed the webview to be hosted in a new `vscode-webview` protocol. We've tried to make this change transparent but it looks like your extension had some advanced usages that we couldn't migrate automatically.

A few quick recommendations:

- Try not to hardcode `vscode-webview-resource` or `vscode-resource` anywhere, since this will not work when VS Code is run in web environments. Instead try using `cspSource`  to get this value if you need it

- Also avoid referencing `vscode-webview:` and instead try getting this information from the webview itself if possible. On web, webviews are hosted inside normal iframes that use a completely different origin

I'll try debugging the extension again with those changes to see if I can figure out what is going on

**Comment by mjbvz** [comment]
@mliao95 Any updates from this on your side? Still blocked?

**Comment by mliao95** [comment]
Still blocked on my end.  I've had my hands tied on some other work, but I plan on working on this tomorrow and next week.  I saw that the breaking change is now in stable, so the extension is currently broken.  Have you figured anything out on your end?

**Comment by robpaveza** [comment]
Here is a minimal repro that exhibits behavior that was functional in 1.46 but fails in 1.47, in the same way that the DevTools one fails to load. I scaffolded this using `yo code`.

https://github.com/robpaveza/vscode-webview-load-problem

**Comment by vidorteg** [comment]
Hi @mjbvz, I'm the other developer working on the extension. I just tested this in VSCode stable (1.47.2) and on insiders(1.48-insider) with the suggestions you mentioned

https://github.com/microsoft/vscode-edge-devtools/commit/31d6c8908ba7be70ecb6b8f1058abc6dada387c3

 and it:

X **broken** on stable
+ **works** on insiders

Stable:
<img width="1200" alt="stable" src="https://user-images.githubusercontent.com/43078681/87731034-512e0a80-c77e-11ea-966c-7694253d25ef.PNG">

Insiders:
<img width="1210" alt="insiders" src="https://user-images.githubusercontent.com/43078681/87731043-55f2be80-c77e-11ea-9cfa-c8eef0b3ac09.PNG">

As a side note, although it works on insider the performance was significantly degraded, before loading was almost immediate now it takes at least 1s (same code)

**Comment by robpaveza** [comment]
For my minimal repro, it looks like the fundamental problem was this:

```
        // Before: 
        const iframeSrc = panel.webview.asWebviewUri(onDiskPath);
        const iframeUri = iframeSrc.with({ scheme: 'vscode-resource' });
        // After:
       const iframeUri = panel.webview.asWebviewUri(onDiskPath);
 ```

We don't know where the `vscode-resource` piece of this was introduced, but as I say, that is what worked correctly in 1.46 and fails in 1.47.

Using the `panel.webview.cspSource`  seems to allow scripts to run just fine as well.

**Comment by robpaveza** [comment]
The branch `upgrade-path` in my minimal repro seems to work fine:

https://github.com/robpaveza/vscode-webview-load-problem/pull/1/files

**Comment by mjbvz** [comment]
Ah, so that would explain the problem. Does that unblock the real edge extension as well?

Just to summarize the problem, in the minimal repo extension, the webview url ends up being incorrectly rewritten to:

```
vscode-webview-resource://cecd63e7-3d7c-42d2-a051-ec82439a34b7/cecd63e7-3d7c-42d2-a051-ec82439a34b7/file///Users/matb/projects/vscode-webview-load-problem/frontend/index.html
```

Notice that the uuid is duplicated here. The proper url from the `upgrade-path` branch instead looks like:

```
vscode-webview-resource://26e45949-c9d4-40a4-b359-5b5cbefe9bde/file///Users/matb/projects/vscode-webview-load-problem/out/host/main.js
```

To avoid breaking extensions, we try to automatically rewrite any old `vscode-resource` urls we see in the html source given to a webview. By calling `asWebviewUri` and then converting the scheme to `vscode-resource`, VS Code will end up  rewriting the url twice in this case.

As you discovered, the fix is to just call `asWebviewUri`. This should handle all the url modifications for you

**Comment by vidorteg** [comment]
No, the extension is still broken on stable but working on 1.48-insiders. (details in my post above)

**Comment by mjbvz** [comment]
Can you try to  create a minimal repo for that as well. The `update-path` branch of the current example works fine for me in insiders and stable VS Code

**Comment by vidorteg** [comment]
Hi @mjbvz  thanks for the quick follow up. 

> As you discovered, the fix is to just call asWebviewUri
We already do this on the DevTools extension.

> - Try not to hardcode vscode-webview-resource or vscode-resource anywhere, since this will not work when VS Code is run in web environments. Instead try using cspSource to get this value if you need it
Same, this is already done.

> - Also avoid referencing vscode-webview: and instead try getting this information from the webview itself if possible. On web, webviews are hosted inside normal iframes that use a completely different origin
We do this with one exception: third_party code (Devtools code).
Devtools uses `script-src self` for content-policy and to make it work we replace it with `script-src vscode-webview-resource: self`, We do this as I'm not sure how to use the webview object as it lives in the extension(host) and the Devtools code is part of the code rendered in the webview.

> Can you try to create a minimal repo for that as well. 
I would but I'm not sure what is happening, there are no errors messages on either the view or the host. And the same code renders on Insiders(1.48-insiders) but is broken now on stable(1.47), more details on the screenshots below.

**Comment by vidorteg** [comment]
Hi @mjbvz do you have any news on this? Is there info we can provide to help?

**Comment by mjbvz** [comment]
Closing this since it seems fixed in VS Code 1.48 (current insiders builds). I just tested again on Windows with the latest VS Code insiders and confirmed the extension does appear to be working

Please let me know if you see this regress against before 1.48 is released

**Comment by billie-spektor** [comment]
Hi @mjbvz 

The extension seem to be broken again in the current insiders build:

![image](https://user-images.githubusercontent.com/69010225/88954898-61c0a380-d29b-11ea-986a-b459a490f918.png)

**Comment by mliao95** [comment]
Confirms that it doesn't work on the most recent version of insiders.

Here is a working version on insiders from a week ago:
![image](https://user-images.githubusercontent.com/14304598/88975468-155f6e80-d26f-11ea-8240-425618975567.png)

**Comment by mliao95** [comment]
@mattbierner I've narrowed down the regression to this commit: https://github.com/microsoft/vscode/commit/86c04f72be7cb690f0304db4c4d4313abd5701c1

I verified that our extension works with the commit right before this and breaks on this commit.  Furthermore, I've verified that our extension will work if we revert this commit at the tip of vscode master.  It seems that the code introduced to parse special characters in the requestUri.path is what is breaking our styling scripts.

**Comment by mjbvz** [comment]
Thanks for the investigation! Any idea which uri in your extension commit this commit breaks?

**Comment by mliao95** [comment]
I'm seeing these errors when running our extension:

`[16492:0803/125757.903:INFO:CONSOLE(96)] "[Embedded Page] Empty response arrived for script 'vscode-webview-resource://8f4c1578-a5ca-47b7-8ecf-63a0dce79eea/file/c%3A/Users/milia/.vscode-oss-dev/extensions/ms-edgedevtools.vscode-edge-devtools-1.0.8/out/tools/front_end/inline_editor/inline_editor_module.js'", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
Webview: Cannot load resource outside of protocol root
Webview: Cannot load resource outside of protocol root
[16492:0803/125757.940:INFO:CONSOLE(96)] "[Embedded Page] Empty response arrived for script 'vscode-webview-resource://8f4c1578-a5ca-47b7-8ecf-63a0dce79eea/file/c%3A/Users/milia/.vscode-oss-dev/extensions/ms-edgedevtools.vscode-edge-devtools-1.0.8/out/tools/front_end/color_picker/color_picker_module.js'", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
[16492:0803/125757.942:INFO:CONSOLE(96)] "[Embedded Page] Empty response arrived for script 'vscode-webview-resource://8f4c1578-a5ca-47b7-8ecf-63a0dce79eea/file/c%3A/Users/milia/.vscode-oss-dev/extensions/ms-edgedevtools.vscode-edge-devtools-1.0.8/out/tools/front_end/help/help_module.js'", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
Webview: Cannot load resource outside of protocol root
[16492:0803/125758.098:INFO:CONSOLE(96)] "[Embedded Page] Empty response arrived for script 'vscode-webview-resource://8f4c1578-a5ca-47b7-8ecf-63a0dce79eea/file/c%3A/Users/milia/.vscode-oss-dev/extensions/ms-edgedevtools.vscode-edge-devtools-1.0.8/out/tools/front_end/object_ui/object_ui_module.js'", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
Webview: Cannot load resource outside of protocol root
[16492:0803/125758.208:INFO:CONSOLE(96)] "[Embedded Page] Empty response arrived for script 'vscode-webview-resource://8f4c1578-a5ca-47b7-8ecf-63a0dce79eea/file/c%3A/Users/milia/.vscode-oss-dev/extensions/ms-edgedevtools.vscode-edge-devtools-1.0.8/out/tools/front_end/event_listeners/event_listeners_module.js'", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
Webview: Cannot load resource outside of protocol root
[16492:0803/125758.514:INFO:CONSOLE(96)] "[Embedded Page] Empty response arrived for script 'vscode-webview-resource://8f4c1578-a5ca-47b7-8ecf-63a0dce79eea/file/c%3A/Users/milia/.vscode-oss-dev/extensions/ms-edgedevtools.vscode-edge-devtools-1.0.8/out/tools/front_end/elements/elements_module.js'", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
[16492:0803/125758.518:INFO:CONSOLE(96)] "[Embedded Page] elements/elementsPanel.css not preloaded. Check module.json", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
[16492:0803/125758.529:INFO:CONSOLE(96)] "[Embedded Page] elements/breadcrumbs.css not preloaded. Check module.json", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
[16492:0803/125758.530:INFO:CONSOLE(96)] "[Embedded Page] elements/stylesSidebarPane.css not preloaded. Check module.json", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
[16492:0803/125758.540:INFO:CONSOLE(96)] "[Embedded Page] inline_editor/swatchPopover.css not preloaded. Check module.json", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
[16492:0803/125758.555:INFO:CONSOLE(96)] "[Embedded Page] elements/computedStyleSidebarPane.css not preloaded. Check module.json", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
[16492:0803/125758.582:INFO:CONSOLE(96)] "[Embedded Page] elements/computedStyleWidgetTree.css not preloaded. Check module.json", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
[16492:0803/125758.586:INFO:CONSOLE(96)] "[Embedded Page] elements/platformFontsWidget.css not preloaded. Check module.json", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
[16492:0803/125758.621:INFO:CONSOLE(96)] "[Embedded Page] elements/metricsSidebarPane.css not preloaded. Check module.json", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
[16492:0803/125758.629:INFO:CONSOLE(96)] "[Embedded Page] elements/elementsTreeOutline.css not preloaded. Check module.json", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
[16492:0803/125758.639:INFO:CONSOLE(96)] "[Embedded Page] elements/elementStatePaneWidget.css not preloaded. Check module.json", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
[16492:0803/125758.661:INFO:CONSOLE(96)] "[Embedded Page] elements/classesPaneWidget.css not preloaded. Check module.json", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
[16492:0803/125758.898:INFO:CONSOLE(96)] "[Embedded Page] elements/stylesSectionTree.css not preloaded. Check module.json", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
[16492:0803/125758.908:INFO:CONSOLE(96)] "[Embedded Page] elements/stylesSectionTree.css not preloaded. Check module.json", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
[16492:0803/125759.023:INFO:CONSOLE(96)] "[Embedded Page] elements/stylesSectionTree.css not preloaded. Check module.json", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)
[16492:0803/125759.025:INFO:CONSOLE(96)] "[Embedded Page] elements/stylesSectionTree.css not preloaded. Check module.json", source: file:///C:/Users/milia/vscode/out/vs/workbench/contrib/webview/electron-browser/webviewElement.js (96)`

It seems pretty relevant as it does not appear on the working version where the offending commit is reverted and these errors pop up if the extension's styles are broken.  Does this help narrow down what the issue might be?

**Comment by mliao95** [comment]
Update:
For a broken URL, 

the original looks like:
vscode-webview-resource://8f4c1578-a5ca-47b7-8ecf-63a0dce79eea/file/c%3A/Users/milia/.vscode-oss-dev/extensions/ms-edgedevtools.vscode-edge-devtools-1.0.8/out/tools/front_end/inline_editor/inline_editor_module.js 

The broken URL looks like:
vscode-webview-resource%3A//8f4c1578-a5ca-47b7-8ecf-63a0dce79eea/file/c%253A/Users/milia/.vscode-oss-dev/extensions/ms-edgedevtools.vscode-edge-devtools-1.0.8/out/tools/front_end/inline_editor/inline_editor_module.js

The differences are:
- the colon at : / /
- c%3A/ vs /c%253A

**Comment by mliao95** [comment]
Additionally, when looking at the output of the resourceURI.path with and without the offending change, we get this:

========Working Version===============
Uri {
  scheme: 'file',
  authority: '',
  path: '/c:/Users/milia/.vscode-oss-dev/extensions/ms-edgedevtools.vscode-edge-devtools-1.0.8/out/tools/front_end/inline_editor/inline_editor_module.js',
  query: '',
  fragment: '',
  _formatted: null,
  _fsPath: null
}


=====Broken Version========
Uri {
  scheme: 'file',
  authority: 'c:',
  path: '/Users/milia/.vscode-oss-dev/extensions/ms-edgedevtools.vscode-edge-devtools-1.0.8/out/tools/front_end/inline_editor/inline_editor_module.js',
  query: '',
  fragment: '',
  _formatted: null,
  _fsPath: null
}

The difference here is the "C:/" at the beginning of the working version which is not present in the broken version.  We get the same issue with other files but this is just one of a handful of examples.

**Comment by vidorteg** [comment]
confirmed the fix. Thanks for the help @mjbvz
